module/zfs/metaslab.c

```diff
@@ -1585,10 +1610,16 @@ metaslab_block_picker(range_tree_t *rt, uint64_t *cursor, uint64_t size,
 	while (rs != NULL && (rs_get_start(rs, rt) - first_found <=
 	    max_search || count_searched < metaslab_min_search_count)) {
 		uint64_t offset = rs_get_start(rs, rt);
+		vdev_t *vd = msp->ms_group->mg_vd;
+
+		if (vd->vdev_ops == &vdev_draid_ops)
+			offset = metaslab_draid_offset(vd, offset, size);
```

Comment:
RAIDZ and DRAID both have requirements on how much space is allocated for a given `psize`.  I'm curious why we have this special case code for draid but not for raidz.  Is this a better way of doing it and should we add raidz-specific code to metaslab_block_picker() as well?

Reply:
There's one additional restriction which applies to draid but not raidz.  The allocation cannot span redundancy groups so we need some extra logic to advance to the start of the next group when we detect this case.

Reply:
I re-read the comment in vdev_draid.c but I'm still not sure what a "group" is.  I think there should be a concept of the amount of data+parity that "goes together", and we have to allocate in units of this.  For example with DRAID2 and the default 8 data disks (+ 2 parity) and 4K sectors (ashift=12), we have to allocate units of 32K of data + 8K of parity = 40KB.  How does this relate to a "group"?

> The allocation cannot span redundancy groups

So I guess that a "group" must be multiple allocation units, since an allocation can be multiple allocation units?

Are the parity sectors "counted" in the draid vdev's offsets?  i.e. when allocating one minimum unit in the above example, does the metaslab/spacemap code indicate that we have allocated 32K, or 40K?

RAIDZ has a similar requirement, to allocate in units of `P+1`.  The parity is included in the RAIDZ vdev offsets and in the allocation.

Reply:
I've update the comment and ASCII art in vdev_draid.c to try and better explain.